### ISSUE TITLE
Object Lifecycle

### ISSUE THREAD

**Issue by ozkatz** [treeverse/lakeFS#1566]

**Comment by guy-har** [comment]
Data retention breaks into two main issues - Lifecycle and Garbage collection

<h1>Life cycle</h1>
<h2>User story</h2>
Users should be able to specify a retention period for data stored in lakeFS. We can do this in a reproducibility period. These are defined as a set of policies defined by the user using a combination of paths and time periods.

<h2>Requirements</h2>

- User will be able to configure a retention policy supporting filtering by prefix
- Once retention job is done, lakeFS expired objects will be expired in S3 as well 
Get request for an expired object would return or 409.
- Objects outside of the underlying storage should not be deleted.

<h2>Proposal</h2>

- Objects’ physical address in the underlying storage will be the same path as their logical path, with a UUID added as a suffix.
For example, the file: lakefs://repo/master/a/b/c/filename will be saved under s3://bucket/a/b/c/filename/8abb1f1d-e33b-4487-9cad-d2f86eb26b92.
This way, the user could use AWS S3 lifecycle policy with prefixes to delete them.
- When uploading an object, a new file is saved under a new name, regardless of whether it already exists. In particular, it gets a new timestamp. If the added file is identical to the existing one, the diff will show that the change is metadata only. 
- If the user commits the change, the file appears on the new commit with the new mtime. The mtime on the previous commit will stay as before. 
- Retention will be performed according to the mtime on each commit. That is, the file may be “Gone” from the first commit, while existing on the second commit.
- Whether or not the mtime becomes part of the identity is an implementation detail. However, it makes sense in this solution that mtime in Graveler will be taken from the storage.

<h2>Notes</h2>

- Copy operation in storage adapters will have to do physical copy.
- Retention could not work with dedup. However, if the same file is saved twice under the same path, we may still deduplicate it.
- We might want to support a migration script to change all existing objects to the new path.
- When we implement the Garbage Collection feature, the user will be able to revert a change, causing the added file to be deleted - to avoid duplication of the same file.


<h1> Garbage collection</h1>

<h2>User story</h2>
Users should be able specify the period of time objects will be available after deleted.

TBD

**Comment by guy-har** [comment]
#1935 Raised issues according to 
1. Slowing down the copy operation
2. deduplication will not be supported anymore

As part of conversations regarding those issues we decided to stay with the current implementation for the following reasons:
1. `meta-copy` would cause race condition when triggering meta-copy while committed clean up is running 
2. Deduplication would cause Complication in lifecycle, objects that should be removed as part of lifecycle may still be referenced in other places
3. We noticed that `copy` commands are done significantly less than `put` commands 
4. The `copy` commands that we did see were always followed by remove, meaning they will be rename operations in the future when those are supported by S3.

**Comment by arielshaqed** [comment]
> #1935 Raised issues according to
> 
>     1. Slowing down the copy operation
> 
>     2. deduplication will not be supported anymore

[These 2](https://github.com/treeverse/lakeFS/issues/1935#issuecomment-838032858) also appear there: 
3. Retention must be done by branch (this user-visible and restricts possible user stories)
4. Backwards compatibility: existing installations don't get retention.

> 
> 
> As part of conversations regarding those issues we decided to stay with the current implementation for the following reasons:
> 
>     1. `meta-copy` would cause race condition when triggering meta-copy while committed clean up is running

I am not sure, that sounds like an implementation or a specification bug.  IIUC such a race exists in all proposed solutions.  E.g. the deleted objects cleanup code runs, I run the *new* copy code, my copy breaks down.  I don't think we care in either place.

>     2. Deduplication would cause Complication in lifecycle, objects that should be removed as part of lifecycle may still be referenced in other places
> 
>     3. We noticed that `copy` commands are done significantly less than `put` commands
> 
>     4. The `copy` commands that we did see were always followed by remove, meaning they will be rename operations in the future when those are supported by S3.

If (4) is true then we *definitely* need to support atomic rename ASAP: many Spark installations "rename" nonatomically by copy-then-delete IIUC.  So today the window for racing is the time for metadata copy plus starting the delete operation.  With content copying, we need to add the time for data copy: it is much longer, and the size of the race window depends on the content length.

**Comment by guy-har** [comment]
As a result of the concerns raised due to physical copy and deduplication we will not implement the object lifecycle as suggested above.

We still need to handle deduplication some way though, the main problem with deduplication is that it could be from any point at time.
I am listing here some options with their pros and cons:
1. Support all deduplication - every time we run the lifecycle job we will validate that addresses could be removed 
	pros: 	Deduplication is protected for ever, Very easy to understand
	cons: 	gets extremely expensive over time
2. Combine with GC (Garbage Collection) process - When running lifecycle together with GC we could run lifecycle only on GC `Active commits` 
	pros: 	Fixed commit history time range 
	cons:  	Can't use without lifecycle, Might be confusing
3. Don't support deduplications - Remove all addresses that should be removed via lifecycle, deduplications will be lost
	pros: 	Fast, "Stronger" for compliance rules
	cons: 	Confusing and error prone, No way to "guard" deduplications

On my opinion, we should start with option 1 and 2 combined, that is, if GC is configured work with GC otherwise check with all commits.

**Comment by guy-har** [comment]
Opened a new issue (#2365) for object lifecycle. 
most of the information here is not relevant anymore.
